tree-wide: port debpython to termux

robertkirkman opened the pull request:
- Fixes https://github.com/termux/termux-packages/issues/25921

- Fixes https://github.com/termux/termux-packages/issues/21658

- Fixes https://github.com/termux/termux-packages/issues/26687

- Fixes https://github.com/termux/termux-packages/issues/25621

#### What is debpython?

- debpython is the commands `py3compile` and `py3clean` from Debian. I
  am calling them that because a large chunk of their source code is
  found inside a folder inside Debian's source code named "debpython"
  - https://salsa.debian.org/cpython-team/python3-defaults/-/tree/5348f704668c0b6c360b6c6fb10153b9c2898af5

- rather than packaging `.pyc` files into packages, `py3compile` and
  `py3clean` can be called from `postinst` and `prerm` scripts to
  generate all `.pyc` for the `.py` files in the package immediately
  after the package is installed, and remove all `.pyc` files immediately
  before uninstalling the package, respectively

- fixes the error `trying to overwrite '/data/data/com.termux/files/usr/lib/python3.12/__pycache__/cProfile.cpython-312.pyc'` when packages were built on-device, but at the same time, also:

- prevents the warnings `dpkg: warning: while removing python, directory '/data/data/com.termux/files/usr/lib/python3.12/site-packages' not empty so not removed` as long as no packages were installed using `pip`

- The `termux_step_create_python_debscripts.sh` can configure work on debpython (i.e. its `py3copile` and `py3clean` commands) from the glibc package `python-glibc`, if some glibc package is being compiled.

- New variables have been implemented:
  - `TERMUX_PYTHON_CROSSENV_BUILDHOME` - location of crossenv's python build libraries.
  - `TERMUX_PKG_PYTHON_RUNTIME_DEPS` - configures the installation of the python modules via pip3 in the pkg's debscripts. If not configured in the package, it will use the value from `TERMUX_PKG_PYTHON_TARGET_DEPS`. If the variable is set to `false`, then the customization of installing python modules will be disabled, even if the `TERMUX_PKG_PYTHON_TARGET_DEPS` variable is set in the package.
  - `TERMUX_SUBPKG_PYTHON_RUNTIME_DEPS` - configures the installation of the python modules via pip3 in the subpkg's debscripts.

- Implemented reconfiguration of prefixes in python module `sysconfig` and setting in `TERMUX_PYTHON_CROSSENV_BUILDHOME`, so that python modules from crossenv building can specify system paths of termux for correct compilation.

- Added automatic addition of `python-glibc{-glibc}` dependency when using the `TERMUX_PKG_PYTHON_RUNTIME_DEPS` (for pkg; will be disabled, i.e. will not be added, if the variable is set to `false`) or `TERMUX_SUBPKG_PYTHON_RUNTIME_DEPS` (for subpkg) value.

#### How to add a new Python package after this?

Everything is the same, except, now, this block is no longer necessary in `build.sh`.

```bash
termux_step_create_debscripts() {
	cat <<- EOF > ./postinst
	#!$TERMUX_PREFIX/bin/sh
	echo "Installing dependencies through pip..."
	pip3 install ${TERMUX_PKG_PYTHON_TARGET_DEPS//, / }
	EOF
}
```

- Instead, `scripts/build/termux_step_create_python_debscripts.sh` can now detect the presence of `pip` package lists in `$TERMUX_PKG_PYTHON_TARGET_DEPS`, `$TERMUX_SUBPKG_PYTHON_TARGET_DEPS`, and the `METADATA` file of the Python package if it exists, and automatically insert them as a block into the `postinst` script for all relevant packages.
- `$TERMUX_PKG_PYTHON_TARGET_DEPS` is used for `pip` dependencies that are both on-device build-time and on-device run-time dependencies, and `$TERMUX_PKG_PYTHON_RUNTIME_DEPS` is used for runtime-only `pip` dependencies. `$TERMUX_PKG_PYTHON_RUNTIME_DEPS` overrides `$TERMUX_PKG_PYTHON_TARGET_DEPS` for runtime dependencies,
  - i.e. if `TERMUX_PKG_PYTHON_RUNTIME_DEPS` is not specified, but `TERMUX_PKG_PYTHON_TARGET_DEPS` is, then `TERMUX_PKG_PYTHON_TARGET_DEPS` will be used as both on-device build and on-device runtime dependencies,
  - but if `TERMUX_PKG_PYTHON_RUNTIME_DEPS` is specified, then `TERMUX_PKG_PYTHON_TARGET_DEPS`, if specified, is used only for on-device build-time dependencies.
- If `python-pip` is not already in the dependencies of any package that needs it, the build will fail with an error instructing maintainers to add `python-pip` to the dependencies of the package that needs it.

licy183 commented:
This folder should only be excluded only when performing on-device build. In some packages, like python, this content in this folder is managed by apt. It is quite annoying to see some warnings like `xxx directory is not empty`.

robertkirkman commented:
> This folder should only be excluded only when performing on-device build. In some packages, like python, this content in this folder is managed by apt

Should the content in these folders remain managed by apt in any case?

In Debian, I believe this folders are **not** managed by `apt`.

https://packages.debian.org/sid/amd64/libpython3.12-minimal/filelist

In Arch Linux, I believe these folders **are** managed by `pacman`.

https://archlinux.org/packages/core/x86_64/python/ (file list)

Fedora also ships them

https://packages.fedoraproject.org/pkgs/python3.12/python3-libs/fedora-40.html

it seems to me like it is optional whether to include them in the package, because I think Python can automatically generate them, but this problem is very difficult because, I think Arch Linux and Fedora might ship them for performance.

At least this does not break Python, I think, because 

> It is quite annoying to see some warnings like `xxx directory is not empty`.

I do not think there is any way to get rid of this warning, because on Termux `python-pip` generates and stores the `__pycache__` folders into the `site-packages` folder in the same `$PREFIX/lib/python3.12/` folder where the `$PREFIX/lib/zoneinfo/_zoneinfo.py` files are, etc.

Example: warning always reproducible with current packages
```bash
pkg install python-pip
pip install setuptools
pkg remove python
```

```
Removing python-pip (25.0.1) ...
Removing python-ensurepip-wheels (3.12.9) ...
Removing python (3.12.9) ...
dpkg: warning: while removing python, directory '/data/data/com.termux/files/usr/lib/python3.12/site-packages' not empty so not removed
Processing triggers for man (1.14.6-1) ...
```

That happens without this change.

Can you think of any way to:

- prevent the `__pycache__` folders from being included in the package during on-device `build-package.sh`, to fix `trying to overwrite '/data/data/com.termux/files/usr/lib/python3.12/__pycache__/cProfile.cpython-312.pyc', which is also in package python 3.12.9`
- but still make sure `__pycache__` folders are included in the package _only_ when they actually belong to the package currently being built, so they can be shipped for performance boost?

For the record: I **do** notice the performance loss at startup of `python` package built from this change - very slightly.

licy183 commented:
In Debain, they have a postinst script to compile the modules when new packages are installed, and a prerm script to clean up `__pycache__` dir and *.pyc files. I have done the same cleanup logic in pypy3 on Termux.

See `/usr/share/python3/runtime.d/public_modules.rt{install,remove}` in Debain.

robertkirkman commented:
> In Debain, they have a postinst script to compile the modules when new packages are installed, and a prerm script to clean up `__pycache__` dir and *.pyc files. I have done the same cleanup logic in pypy3 on Termux.
> 
> See `/usr/share/python3/runtime.d/public_modules.rt{install,remove}` in Debain.

ok I did it, is this what you meant?

robertkirkman commented:
@Maxython 
Ok, to continue the discussion here,

now I have:

- unified almost all python debscript creation into a single function, removing the duplicated code involving `$TERMUX_PKG_PYTHON_TARGET_DEPS` from many packages and also making it not necessary to explicitly call `termux_step_create_python_debscripts()` from any function
- tried to make this debpython compatible with `pacman` by making it try to use the `pacman -Qql` command.

however, I am not sure whether the `pacman -Qql` command is completely compatible to the point that it will work in all cases without problems. `pacman -Qql` does print equivalent data in very similar formatting to `dpkg -L`, but it is **not _100% identical formatting_** to `dpkg -L`, so I am not sure whether it will work perfectly without a lot of testing.

TomJo2000 reviewed (approved):
Thanks for putting in the effort.
Looks like I accidentally created a merge conflict for you with one of my earlier PRs, sorry about that.

robertkirkman commented:
That's ok,
something I have noticed is that, licy183 told me somewhere else that `TERMUX_PKG_REVISION` should not be incremented for PRs that are excluded from CI. The reason I excluded the PR from CI is because I previously learned [here](https://github.com/termux/termux-packages/pull/21835#issuecomment-2447031658) that if too many packages are built in a single PR, sometimes the behavior will diverge from the behavior of the same large build command run in the docker builder locally.

However, the reason why many packages are marked with revision-bumps in this is because, that is marking all packages that a test I did indicates will have **changed `control.tar.xz` contents** if they are rebuilt after this PR is merged.

Because of that, I think probably what I need to do is split this into several PRs one after another that do not have CI disabled.

Maxython commented:
The following packages need to be recompiled using debpython:

- [ ] packages/borgbackup
- [ ] packages/electrum
- [ ] packages/hash-slinger
- [ ] packages/lv2
- [ ] packages/manim
- [ ] packages/nala
- [ ] packages/python-cryptography
- [ ] packages/python-tldp
- [ ] packages/python-xlib
- [ ] packages/rdiff-backup
- [ ] packages/rdircd
- [ ] packages/sabnzbd
- [ ] packages/srt2vobsub
- [ ] packages/termux-gui-package
- [ ] packages/termux-gui-pm
- [ ] packages/ytui-music
- [ ] x11-packages/orca
- [ ] x11-packages/pyqt5

TomJo2000 commented:
> The following packages need to be recompiled using debpython:

I would suggest either opening a staging PR for this purpose, or added the commits for rebuilding those packages to this PR.

TomJo2000 commented:
> It is planned to rebuild these packages in each PRs. If rebuild all the packages in one PR (for example on this one), there are risks that can break the compilation.

I'm not sure I understand that concern.

Maxython commented:
> I'm not sure I understand that concern.

Well, there are quite a lot of packages that need to be rebuilt. And one of the risks I see is that the packages may not have time to be built, i.e. their compilation may take more than six hours.

TomJo2000 commented:
Those 18 packages should get nowhere near 6 hours.
The largest one there is `pyqt5` which took ~20 minutes the last time it was built.

robertkirkman commented on packages/nala/build.sh:
```diff
@@ -42,7 +43,5 @@ termux_step_create_debscripts() {
 	mkdir -p $TERMUX_PREFIX/var/log/nala
 	mkdir -p $TERMUX_PREFIX/var/lock
-	echo "Installing dependencies through pip..."
-	pip3 install nala python-debian
```
Correct, I do understand that for packages that use `pip3` to install **a `pip` equivalent of the same package as the current package**,

so for packages that do that, like `matplotlib` for example, I have intentionally not edited them in this PR, since they should not be changed.

The way this is currently for `nala` in this PR is incorrect and needs to be adjusted. I believe this must be a typo, and it was meant to be a different change, `pip3 install nala python-debian` -> `pip3 install nala`.

Thank you for pointing this out. The commit history of this PR has shifted and that change is now labeled under with Maxython's commit, but I believe the original responsibility for this typo 

> I think it is better to use `pip3 install package_name` to let pip handle the version of dependencies automatically rather than use `TERMUX_PKG_PYTHON_TARGET_DEPS`.

The issue with doing it that way for _all_ packages is that a few packages are not on PyPi, but have runtime dependencies that _are_ on PyPi, so `pip3 install package_name` would not work in those situations. Therefore, there has to be a mixture of both kinds of python debscripts.

robertkirkman replied:
I have added a fix for this typo into the current version of this PR

Maxython replied:
> I think it is better to use pip3 install package_name to let pip handle the version of dependencies automatically rather than use TERMUX_PKG_PYTHON_TARGET_DEPS.

Could you explain in more detail why it is better to leave the `nala` module installation setting in `termux_step_create_debscripts`, since when configuring the `nala` module in the `TERMUX_PKG_PYTHON_TARGET_DEPS` variable, its installation in debscripts will be automatically added?

robertkirkman replied:
In the case of `nala` specifically, it could be done and still work, yes. I have just tested that now. 

However, I just didn't remove the manual debscript from the packages which did that (install a package from PyPi with a similar name to the package name) because I thought there would be a possibility of an error with some of them during on-device builds, if the package from PyPi that has a similar name to the package name were placed into `TERMUX_PKG_PYTHON_TARGET_DEPS`.

If you would like, I guess we can test and try to revise all existing main packages (not subpackages) that have `pip` debscripts, and see whether they can all be removed definitively without causing an error anywhere, and maybe if some cause errors, only the ones that can be removed without causing errors could be removed.

licy183 replied:
> Could you explain in more detail why it is better to leave the `nala` module installation setting in `termux_step_create_debscripts`, since when configuring the `nala` module in the `TERMUX_PKG_PYTHON_TARGET_DEPS` variable, its installation in debscripts will be automatically added?

`nala` should depend on specified version of `httpx` (`httpx<0.24.0,>=0.23.0`). If `postinst` has `pip3 install anyio httpx jsbeautifier pexpect python-debian rich tomli typer typing-extensions` and user has installed `httpx>=0.24.0` themselves (or by the `postinst` script of other packages), `pip` will not resolve the dependencies automatically and will show an ERROR message every time `pip` is executed. So, if one package is managed by `pip`, it is better to use `pip install package_name` to let pip resolve these dependencies rather than `pip install some_deps`.


This error is reported in #22686:

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
nala 0.15.4 requires anyio<4.0.0,>=3.6.1, but you have anyio 4.7.0 which is incompatible.
nala 0.15.4 requires httpx<0.24.0,>=0.23.0, but you have httpx 0.28.1 which is incompatible.
nala 0.15.4 requires rich<13.0.0,>=12.4.4, but you have rich 13.9.4 which is incompatible.
nala 0.15.4 requires typer<0.5.0,>=0.4.1, but you have typer 0.15.1 which is incompatible.
```

robertkirkman replied:
ok that makes a lot of sense thank you!
I need to make another change here.

robertkirkman replied:
The version you just added @Maxython might be sufficient to prevent all errors in `nala`, because of it putting all dependencies in one `pip` command. however, for the other packages, it might not be as easy. let's test all of them to check if any have errors like this. I will check them when I can

Maxython replied:
I have reconfigured the method for configuring the installation of python modules in debscripts, and now you can use the variable `TERMUX_PKG_PYTHON_SCRIPT_DEPS` for this, instead of `TERMUX_PKG_PYTHON_TARGET_DEPS`. That is, now in the `nala` package, the installation of the `nala` module occurs through `TERMUX_PKG_PYTHON_SCRIPT_DEPS`, separately from the modules in `TERMUX_PKG_PYTHON_TARGET_DEPS`

robertkirkman commented on scripts/build/termux_step_create_python_debscripts.sh:
```diff
@@ -0,0 +1,109 @@
+
+	# running py3compile in a package for pacman during a package update
+	if [ "$TERMUX_PACKAGE_FORMAT" = "pacman" ] && ! grep -qs 'post_install' postupg; then
+		echo "post_install" >> postupg
```
> [!NOTE]
> This part is loosely related to this
> - https://github.com/termux/termux-packages/pull/25222
> however, they should not conflict at all with each other even slightly, because the other check checks for a preexisting `postupg` that could be from this check, and does nothing in that case.

Maxython commented on packages/python/debpython.patch:
```diff
@@ -37,34 +33,6 @@ index 39ef978..5899a7a 100644
-index 45b14db..6334661 100644
---- a/debpython/debpython/interpreter.py
-+++ b/debpython/debpython/interpreter.py
-@@ -171,16 +171,16 @@ class Interpreter:
```
Why do you think these changes are not necessary? These changes are made so that the `sitedir` function can return the actual path to python's `site-packages` in Termux. And although this function is not used anywhere in the debpython system, it can be used by users or projects.

robertkirkman replied:
Could you provide an example of how to activate and test this code?

I was interested in finding a way to activate it as well, but I couldn't find one.

robertkirkman replied:
Ok thanks a lot, that's very interesting and I didn't really know that was a meaningful use-case. I will undo the change that messes up those commands.

robertkirkman replied:
I think it is very helpful that there is now a note here that it's possible to just use `from debpython import` and it will work in Python after this.

robertkirkman commented:
Yes so, Tomjo2000, what we're interested in is whether all these packages will install and work successfully on-device after building here, and without showing the message:

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
```

robertkirkman commented:
> `Compatibility with CMake < 3.5 has been removed from CMake.`

After

- #25014 

CMake in all of Termux is now CMake 4, so packages that require CMake 3 cannot build anymore until they receive updates.

robertkirkman commented on scripts/build/termux_step_create_python_debscripts.sh:
```diff
@@ -0,0 +1,112 @@
+	if [[ -n "${SUB_PKG_NAME-}" ]]; then
+		local _package_name="$SUB_PKG_NAME"
+		local _package_massagedir="$SUB_PKG_DIR/massage/$TERMUX_PREFIX"
+		local _package_python_deps="$TERMUX_SUBPKG_PYTHON_SCRIPT_DEPS"
```
These are now the first instances of `TERMUX_SUBPKG_PYTHON_SCRIPT_DEPS` to be found in the repository (`SUBPKG` as opposed to `PKG`),

Also, it will be necessary to make adaptations for many packages in the repository and test them with `TERMUX_PKG_PYTHON_SCRIPT_DEPS`,

had you already started doing that, or would you like me to help you find and change packages to this new variable and test them, @Maxython ?

Maxython replied:
> These are now the first instances of TERMUX_SUBPKG_PYTHON_SCRIPT_DEPS to be found in the repository (SUBPKG as opposed to PKG), so it would be necessary to add this variable to some subpackages, to replace the termux_step_create_subpkg_debscripts() functions in them

I seem to have updated all subpackages and now they all use this variable (`TERMUX_SUBPKG_PYTHON_SCRIPT_DEPS`).

> Also, it will be necessary to make adaptations for many packages in the repository and test them with TERMUX_PKG_PYTHON_SCRIPT_DEPS, or if the name I suggested is chosen, TERMUX_PKG_PYTHON_RUNTIME_DEPS

I will change the `TERMUX_PKG_PYTHON_SCRIPT_DEPS` variable to `TERMUX_PKG_PYTHON_RUNTIME_DEPS` after the packages in the PR are compiled.

robertkirkman commented on packages/matplotlib/build.sh:
```diff
@@ -15,12 +15,14 @@ LICENSE/LICENSE_STIX
+TERMUX_PKG_DEPENDS="freetype, libc++, patchelf, ninja, python, python-contourpy, python-numpy, python-pillow"
 _NUMPY_VERSION=$(. $TERMUX_SCRIPTDIR/packages/python-numpy/build.sh; echo $TERMUX_PKG_VERSION)
 TERMUX_PKG_PYTHON_COMMON_DEPS="build, 'meson-python>=0.13.1', wheel, 'numpy==$_NUMPY_VERSION', 'pybind11>=2.6.0', 'setuptools>=64', 'setuptools_scm>=7'"
+TERMUX_PKG_PYTHON_TARGET_DEPS="matplotlib"
```
There are many other places where this kind of change will be necessary, and I would also like to change the names of all the variables, but I believe Maxython has chosen the original names of the variables, so unfortunately, it might be necessary to reach a consensus on variable names before deciding exactly which changes to make, and where. I have written more information about this in one of my other messages

Maxython commented on scripts/build/termux_step_setup_variables.sh:
```diff
@@ -173,7 +173,9 @@ termux_step_setup_variables() {
 	TERMUX_PKG_PYTHON_TARGET_DEPS="" # python modules to be installed via pip3
```
> I really don't think the design of `TERMUX_PKG_PYTHON_TARGET_DEPS` is correct.
> 
> Currently, `TERMUX_PKG_PYTHON_TARGET_DEPS` is used in two places, one is for the on-device build, before compiling the current python wheel, as a pre-build dependency, and the other is in the `postinst` script, as a runtime dependency. In fact, the dependencies which are necessary for the on-device build are not the same as the runtime dependencies. The design of this variable should be reconsidered.

Technically, using the `TERMUX_PKG_PYTHON_TARGET_DEPS` variable in `postinst` makes sense, because the variable itself is supposed to store python modules that need to be installed in the Termux application (i.e. in the native), and most often the python modules from this variable are needed for compilation (i.e. they are like a dependency for building) and for operation (i.e. they are like a normal dependency) so that python packages can work normally. 

> the names and use of all four variables should be renamed and reorganized into more accurate and more easily manageable categories

The `TERMUX_PKG_PYTHON_SCRIPT_DEPS` variable needs some more thought. Your suggestion to use the `TERMUX_PKG_PYTHON_RUNTIME_DEPS` variable instead of `TERMUX_PKG_PYTHON_SCRIPT_DEPS` looks interesting. As for other python variables (`TERMUX_PKG_PYTHON_BUILD_DEPS`, `TERMUX_PKG_PYTHON_TARGET_DEPS` and `TERMUX_PKG_PYTHON_COMMON_DEPS`), I'm not sure that this is necessary.

robertkirkman replied:
> Technically, using the TERMUX_PKG_PYTHON_TARGET_DEPS variable in postinst makes sense, because the variable itself is supposed to store python modules that need to be installed in the Termux application (i.e. in the native), and most often the python modules from this variable are needed for compilation (i.e. they are like a dependency for building) and for operation (i.e. they are like a normal dependency) so that python packages can work normally.

I Believe that this is not quite 100% correct, because as we have seen, there are certain packages like `nala`, `python-onnxruntime`, where the idea of installing the exact same PyPi dependencies list at build-time and run-time does not work on-device. so, on-device build-time and run-time dependencies must have separate lists in at least some cases, so we need to figure out some way to consistently specify this. In my opinion, four variables is probably a sufficient number to store the necessary set of possible configurations.

> As for other python variables (TERMUX_PKG_PYTHON_BUILD_DEPS, TERMUX_PKG_PYTHON_TARGET_DEPS and TERMUX_PKG_PYTHON_COMMON_DEPS), I'm not sure that this is necessary.

From the first time I saw these variables, I have to admit that I wanted to try to change their names to be easier to understand, but

licy183 replied:
The variable name I proposed at the time was misleading, and it turns out that the variable has been misused because of its name.

robertkirkman replied:
> The variable TERMUX_PKG_PYTHON_TARGET_DEPS plays as the second variable TERMUX_PKG_PYTHON_RUNTIME_DEPS, if the variable TERMUX_PKG_PYTHON_RUNTIME_DEPS is not configured in build.sh.

I am ok with this, this makes sense to me and if this works well when testing, then let's do it that way, yes.

> I think there is sense in the question of renaming them (so that they become more understandable, as you noted). As for the question of reconfiguring their function, as I already said, I think there is no need.

I would be ok with renaming two of the remaining variables to the names I proposed, without changing their functionality.

I think that `TERMUX_PKG_PYTHON_COMMON_DEPS` should be named `TERMUX_PKG_PYTHON_COMMON_BUILD_DEPS` and `TERMUX_PKG_PYTHON_BUILD_DEPS` should be named `TERMUX_PKG_PYTHON_CROSS_BUILD_DEPS`, because these are both groups that are never runtime dependencies and these names more clearly explain the difference between them.

Originally when I first used these variables while editing packages, I was confused about the difference between the variables, and I had to read the code carefully and test the behavior of the variables to find out what really happens.

I was confused whether the `TERMUX_PKG_PYTHON_BUILD_DEPS` variable would make dependencies get installed during on-device builds, and whether `TERMUX_PKG_PYTHON_COMMON_DEPS` variable's list would be installed during package installation or not.

I think if `TERMUX_PKG_PYTHON_CROSS_BUILD_DEPS` and `TERMUX_PKG_PYTHON_COMMON_BUILD_DEPS` are used, they would make it easier for developers from other projects who have not made Termux packages before, to read and understand the purpose of the variable contents without having to read all the code in the `scripts` folder related to these variables.

robertkirkman replied:
I think bulk renaming without changing behavior can be done another time separately, since too much renaming of preexisting variables would make this PR harder to read.

robertkirkman commented:
After some more testing and consideration, I have decided to add some more changes into another temporary commit here,
Maxython, you can read my most recent commit in order to review my newest changes and decide whether you agree with them.

Maxython commented:
LGTM

robertkirkman commented:
I had been testing this and encountering this error during installation of `nala`

```
Errno 2] No such file or directory: '/data/data/com.termux/files/usr/lib/python3.12/site-packages/nala/cache.py'
dpkg: error processing package nala (--configure):
```

since this was technically an error producing a failing result, I created a way to prevent this error and have applied it to the recent commit, and 

This will make `py3compile` skip compiling `.py` files if `dpkg` reports they are in the package, but they end up not being present when `py3compile` runs because of `pip install` deleting those files for some reason if the PyPi equivalent name of the software being packageed is in `TERMUX_PKG_PYTHON_RUNTIME_DEPS`.

After this, I think this PR content is ready, and the commits can be squashed into a single commit. What do you think @Maxython ?

I had some other ideas for additional features or changes that change even more than this, but they are too big for this PR and if I still want them I will make other PRs for them later.

Maxython commented:
> `pip install` deleting those files for some reason if the PyPi equivalent name of the software being packageed is in `TERMUX_PKG_PYTHON_RUNTIME_DEPS`.

It seems that this happens because during the installation of the module, `pip` updates it to the latest version (i.e. this happens because `pip` has the `--upgrade` flag set). And due to this update, the contents of the module's `*.py` files may change (i.e. they may be deleted or new ones may be created).

> After this, I think this PR content is ready, and the commits can be squashed into a single commit. What do you think @Maxython ?

I think we should first reconsider the method for solving this problem you found. Then this PR can be merged with two commits (in one commit the implementation of debpython, in the other recompilation of some packages).

robertkirkman commented:
> I think we should first reconsider the method for solving this problem you found. Then this PR can be merged with two commits (in one commit the implementation of debpython, in the other recompilation of some packages).

Ok sure, sounds great. If you have a specific idea for how to fix the problem of `Errno 2] No such file or directory: '/data/data/com.termux/files/usr/lib/python3.12/site-packages/nala/cache.py'` in a better way, then feel free to force push this branch with any change you want to show, and we can see how that way looks

Maxython commented:
> If you have a specific idea for how to fix the problem of Errno 2] No such file or directory: '/data/data/com.termux/files/usr/lib/python3.12/site-packages/nala/cache.py' in a better way, then feel free to force push this branch with any change you want to show, and we can see how that way looks

The idea for solving the problem: if in the list of modules to install during the installation of packages there is a module whose name matches the name of the package, then the `--upgrade` flag will be disabled.

licy183 commented on scripts/build/termux_step_create_python_debscripts.sh:
```diff
@@ -0,0 +1,116 @@
+	# make this script install them
+	if [[ -n "${_package_python_deps}" ]]; then
+		local upgrade_flag="--upgrade"
+		if [[ " $(tr ' ' '\n' <<< "${_package_python_deps}" | sed "s/'//g; s/</ /g; s/>/ /g; s/=/ /g" | awk '{printf $1 " "}')" =~ " ${_package_name//python-/} " ]]; then
```
The `nala` package on PyPI is not the one packaged by Termux...

https://pypi.org/project/nala/

robertkirkman replied:
I see, but it seems like you added this dependency earlier,

https://github.com/termux/termux-packages/issues/22686

maybe that should be changed, and maybe the dependencies of the Termux `nala` package should be rewritten to be more like the dependencies of the current Debian `nala` package?

https://gitlab.com/volian/nala/-/blob/7784856980039ecb5f1a4449fd03dd569889ec60/debian/control#L33

robertkirkman replied:
Oh ok, so do you mean that `pip install nala` will install PyPi dependencies for the current (_local_) package in `site-packages/nala`, which is different from the PyPi package, but `pip install --upgrade nala` will install the PyPi `nala`, which is different software?

Thank you for explaining this, I really did not know that before. I assumed that `pip` would always treat _all_ package names passed as PyPi packages.

robertkirkman commented:
Ok, I have tested installing the current version on two termux-apt devices with two different CPU architectures, and one clean installed and one not clean installed, and I can confirm there are no more **fatal errors** anywhere in this PR.

However, the one remaining possible problem I see is, I still frequently see these errors:

![Screenshot_20250718-150218_Termux](https://github.com/user-attachments/assets/46fd092d-e820-44bd-acb3-4b348be8511a)

licy183 mentioned these errors first, but unfortunately, I don't know how to fix these.

They are non-fatal errors, because after they show, "Successfully installed" prints anyway.

Also, they do not necessarily produce runtime errors in the python programs, because for example, after this I tested commands like `manim --show-splash init project`, and they appear to work without errors.

On the other hand, I am not sure whether these errors indicate that subtle, hard-to-find edge cases in some programs might appear.

Do you think it is acceptable to allow these errors, or do you know of any way to prevent them?

robertkirkman commented:
I checked and noticed that the dependencies logged as "not installed", do actually appear as installed at the end after all debscripts finish running.

![Screenshot_20250718-153821_Termux](https://github.com/user-attachments/assets/247959af-7fc3-4016-99c0-46f8c65c4a65)

for example, `mapbox-earcut` 1.0.2 is installed after the `apt install` command finishes, but I didn't manually install it.

This suggests the error might be spurious, or also, it could be a coincidence that `mapbox-earcut` was installed by a different debscript somehow, from the one that printed that message. I'm not sure which is the case.

licy183 commented:
> Do you think it is acceptable to allow these errors, or do you know of any way to prevent them?

As I commented before, if you install some package manually (`pip install manim`), `pip` will automatically change these dependencies to correct version.

robertkirkman commented:
Thank you for explaining, I am sorry for being dense about this detail. I think I am beginning to get an idea of what should be done. I will think about this and make a solution.

robertkirkman commented:
After the most recent changes, when I test fully recompiling and reinstalling every single package here at once into a clean Termux installation, **the amount of visible instances of `ERROR: pip's dependency resolver does not currently take into account all the packages that are installed` has not been 100% completley elimitated, but they have been greatly reduced,** and all of the instances remaining seem to occur because `apt` installs all the python package _files_ into `$PREFIX` _before_ ever running any debscripts, then runs the `postinst` scripts one by one for each package, so these messages sometimes print temporarily while not all packages have had their `postinst` scripts run yet - but **I believe that all PyPi dependencies are now installed correctly by the time all `postinst` scripts have finished running.**

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
borgbackup 1.4.1 requires msgpack<=1.1.0,>=1.0.3, which is not installed.
matplotlib 3.10.3 requires cycler>=0.10, which is not installed.
matplotlib 3.10.3 requires fonttools>=4.22.0, which is not installed.
matplotlib 3.10.3 requires kiwisolver>=1.3.1, which is not installed.
matplotlib 3.10.3 requires pyparsing>=2.3.1, which is not installed.
matplotlib 3.10.3 requires python-dateutil>=2.7, which is not installed.
torch 2.6.0 requires filelock, which is not installed.
torch 2.6.0 requires fsspec, which is not installed.
torch 2.6.0 requires jinja2, which is not installed.
torch 2.6.0 requires networkx, which is not installed.
torch 2.6.0 requires setuptools; python_version >= "3.12", which is not installed.
torch 2.6.0 requires typing-extensions>=4.10.0, which is not installed.
torch 2.6.0 requires sympy==1.13.1; python_version >= "3.9", but you have sympy 1.14.0 which is incompatible.
Successfully installed coloredlogs-15.0.1 flatbuffers-25.2.10 humanfriendly-10.0 mpmath-1.3.0 packaging-25.0 protobuf-6.31.1 sympy-1.14.0
Setting up ffmpeg (7.1.1-4) ...
Setting up python-torch (2.6.0-1) ...
Installing dependencies for python-torch through pip...
Requirement already satisfied: torch in /data/data/com.termux/files/usr/lib/python3.12/site-packages (2.6.0)
Collecting filelock (from torch)
  Downloading filelock-3.18.0-py3-none-any.whl.metadata (2.9 kB)
Collecting typing-extensions>=4.10.0 (from torch)
  Downloading typing_extensions-4.14.1-py3-none-any.whl.metadata (3.0 kB)
Collecting setuptools (from torch)
  Downloading setuptools-80.9.0-py3-none-any.whl.metadata (6.6 kB)
Collecting sympy==1.13.1 (from torch)
  Downloading sympy-1.13.1-py3-none-any.whl.metadata (12 kB)
Collecting networkx (from torch)
  Downloading networkx-3.5-py3-none-any.whl.metadata (6.3 kB)
Collecting jinja2 (from torch)
  Downloading jinja2-3.1.6-py3-none-any.whl.metadata (2.9 kB)
Collecting fsspec (from torch)
  Downloading fsspec-2025.7.0-py3-none-any.whl.metadata (12 kB)
Requirement already satisfied: mpmath<1.4,>=1.1.0 in /data/data/com.termux/files/usr/lib/python3.12/site-packages (from sympy==1.13.1->torch) (1.3.0)
```

In particular, I noticed closely that I think all of the dependencies are now being resolved as well as possible for every single package, one by one as the `postinst` scripts for each package that has the `pip`-facing package name installed explicitly without `--upgrade` during `postinst`:

```
torch 2.6.0 requires filelock, which is not installed.
```
...

```
Setting up python-torch (2.6.0-1) ...
Installing dependencies for python-torch through pip...
Requirement already satisfied: torch in /data/data/com.termux/files/usr/lib/python3.12/site-packages (2.6.0)
Collecting filelock (from torch)
  Downloading filelock-3.18.0-py3-none-any.whl.metadata (2.9 kB)
```

@Maxython The amount of changes I have made to do this is quite large, and I am not sure whether you agree with all of the changes. Please let me know if you think everything I have done here is a good idea.

however, after you have finished reviewing my new changes, and any follow-up changes have finished, I believe this PR will be ready.

robertkirkman commented on scripts/buildorder.py:
```diff
@@ -39,7 +39,7 @@ def parse_build_file_dependencies_with_vars(path, vars):
         for line in build_script:
             if line.startswith(vars):
-                dependencies_string = line.split('DEPENDS=')[1]
+                dependencies_string = line.split('DEPS=' if 'DEPS=' in line else 'DEPENDS=')[1]
```
@Maxython I am a little bit confused about this line, and I don't know if I need to change it to match my other recent changes somehow. If it's easy for you to explain, could you explain a little bit about what this line is supposed to do?

robertkirkman reviewed (commented):
EDIT: this comment is an accident, please ignore this comment and read the other one below.
```
          |
          v
```
<img width="1476" height="261" alt="image" src="https://github.com/user-attachments/assets/2dadbf4f-308c-414f-978c-6cb16bbc5b8d" />

robertkirkman commented:
Basically, I need it and it's really inconvenient for me not to have all the necessary changes in one branch, so I've now added **all additional changes that are required to fix**

- https://github.com/termux/termux-packages/issues/21658

This means that after 10 months in development, during which many other fixes have been merged in other PRs that have whittled down more and more of the errors here and there, this PR is now a 100% solution to that issue (`./build-package.sh -I -f mesa` on-device), instead of only part of the solution.

> [!TIP]
> This PR that was recently merged globally fixed a large number of on-device build errors that I had been aware of for years,
> - https://github.com/termux/termux-packages/pull/25627
> 
> and this makes up a portion of the fixes for that issue, but I was unfortunately a bit too unassertive to believe that I could convincingly argue for that change to be made for a long time (which ended up being until NDK r28c broke the existing code further, giving me the opportunity to prove that my approach is better in a high-priority situation)
> 
> - https://github.com/termux/termux-packages/issues/25622

robertkirkman commented:
@hansm629 I believe this PR should now form an implementation of a 100% solution to your October 4 2024 issue,

if you would like to, you can try testing it to just confirm that on your end, by using these commands, which should ideally be **100% successful without any other manual intervention, with no errors**.

```bash
git clone -b exclude-pycache-during-extract-into-massagedir https://github.com/robertkirkman/termux-packages.git
cd termux-packages/
scripts/setup-termux.sh
./build-package.sh -I -f mesa
```

robertkirkman committed:
tree-wide: port debpython to termux

- debpython is the commands `py3compile` and `py3clean` from Debian. I
  am calling them that because a large chunk of their source code is
  found inside a folder inside Debian's source code named "debpython"
  - https://salsa.debian.org/cpython-team/python3-defaults/-/tree/5348f704668c0b6c360b6c6fb10153b9c2898af5

- rather than packaging `.pyc` files into packages, `py3compile` and
  `py3clean` can be called from `postinst` and `prerm` scripts to
  generate all `.pyc` for the `.py` files in the package immediately
  after the package is installed, and remove all `.pyc` files immediately
  before uninstalling the package, respectively

- fixes the error `trying to overwrite '/data/data/com.termux/files/usr/lib/python3.12/__pycache__/cProfile.cpython-312.pyc'` when packages were built on-device, but at the same time, also:

- prevents the warnings `dpkg: warning: while removing python, directory '/data/data/com.termux/files/usr/lib/python3.12/site-packages' not empty so not removed` as long as no packages were installed using `pip`

- The `termux_step_create_python_debscripts.sh` can configure work on debpython (i.e. its `py3copile` and `py3clean` commands) from the glibc package `python-glibc`, if some glibc package is being compiled.

- New variables have been implemented:
- `TERMUX_PYTHON_CROSSENV_BUILDHOME` - location of crossenv's python build libraries.
- `TERMUX_PKG_PYTHON_RUNTIME_DEPS` - configures the installation of the python modules via pip3 in the pkg's debscripts. If not configured in the package, it will use the value from `TERMUX_PKG_PYTHON_TARGET_DEPS`. If the variable is set to `false`, then the customization of installing python modules will be disabled, even if the `TERMUX_PKG_PYTHON_TARGET_DEPS` variable is set in the package.
- `TERMUX_SUBPKG_PYTHON_RUNTIME_DEPS` - configures the installation of the python modules via pip3 in the subpkg's debscripts.

- Implemented reconfiguration of prefixes in python module `sysconfig` and setting in `TERMUX_PYTHON_CROSSENV_BUILDHOME`, so that python modules from crossenv building can specify system paths of termux for correct compilation.

- Added automatic addition of `python-glibc{-glibc}` dependency when using the `TERMUX_PKG_PYTHON_RUNTIME_DEPS` (for pkg; will be disabled, i.e. will not be added, if the variable is set to `false`) or `TERMUX_SUBPKG_PYTHON_RUNTIME_DEPS` (for subpkg) value.

> How to add a new Python package after this?

Everything is the same, except, now, this block is no longer necessary in `build.sh`.

```bash
termux_step_create_debscripts() {
	cat <<- EOF > ./postinst
	#!$TERMUX_PREFIX/bin/sh
	echo "Installing dependencies through pip..."
	pip3 install ${TERMUX_PKG_PYTHON_TARGET_DEPS//, / }
	EOF
}
```

- Instead, `scripts/build/termux_step_create_python_debscripts.sh` can now detect the presence of `pip` package lists in `$TERMUX_PKG_PYTHON_TARGET_DEPS`, `$TERMUX_SUBPKG_PYTHON_TARGET_DEPS`, and the `METADATA` file of the Python package if it exists, and automatically insert them as a block into the `postinst` script for all relevant packages.
- `$TERMUX_PKG_PYTHON_TARGET_DEPS` is used for `pip` dependencies that are both on-device build-time and on-device run-time dependencies, and `$TERMUX_PKG_PYTHON_RUNTIME_DEPS` is used for runtime-only `pip` dependencies. `$TERMUX_PKG_PYTHON_RUNTIME_DEPS` overrides `$TERMUX_PKG_PYTHON_TARGET_DEPS` for runtime dependencies,
  - i.e. if `TERMUX_PKG_PYTHON_RUNTIME_DEPS` is not specified, but `TERMUX_PKG_PYTHON_TARGET_DEPS` is, then `TERMUX_PKG_PYTHON_TARGET_DEPS` will be used as both on-device build and on-device runtime dependencies,
  - but if `TERMUX_PKG_PYTHON_RUNTIME_DEPS` is specified, then `TERMUX_PKG_PYTHON_TARGET_DEPS`, if specified, is used only for on-device build-time dependencies.
- If `python-pip` is not already in the dependencies of any package that needs it, the build will fail with an error instructing maintainers to add `python-pip` to the dependencies of the package that needs it.

Co-authored-by: Maxython <mixython@gmail.com>
